[#20] 디렉터 목록 및 디렉터 상세 조회 기능 구현

tilsong opened the pull request:
## PR 유형

- [ ] 버그 수정
- [x] 기능
- [x] 코드 스타일
- [ ] 리팩토링
- [ ] 빌드
- [ ] CI/CD
- [ ] 문서
- [ ] 기타
  <br>

## 변경사항

### 변경 전
- 디렉터 목록 조회, 디렉터 상세 조회 기능 x
### 변경 후
- 디렉터 목록 조회, 디렉터 상세 조회 기능을 구현했습니다.
- 일부 도메인의 속성을 vo로 변경했습니다. (상세 커밋 참조)

tilsong committed:
[#20] 디렉터 상세 조회 기능 구현 - 상세 조회 기능 구현 및 도메인 관련 코드 리팩토링

tilsong committed:
[#20] 디렉터 상세 조회 기능 구현 - User 도메인 관련 로직 수정

tilsong committed:
[#20] 디렉터 상세 조회 기능 구현 - User 도메인 관련 로직 수정2, 기타(상세 참조)

- Special 도메인 변경: 속성인 specialtyProperty와 description를 하나의 VO로 변경 -> SpecialtyInfo

- Region distance의 기본 단위를 1000(m) 상수로 표현

tilsong committed:
[#20] 디렉터 목록 조회 기능 및 디렉터 상세 조회 기능 구현

- 디렉터 조회 클래스 명 변경: SearchService -> SearchDirectorService

- Controller, Repository에 관련 로직 추가

f-lab-fi commented on src/main/java/com/directors/presentation/user/UserController.java:
```diff
@@ -82,4 +83,19 @@ public ResponseEntity<AuthenticateRegionResponse> authenticateRegion(@RequestBod
         return new ResponseEntity<>(response, HttpStatus.OK);
     }
+
+    @GetMapping("/getDirector/{directorId}")
+    public ResponseEntity<GetDirectorResponse> getDirector(
+            @PathVariable @NotBlank(message = "입력 값이 존재하지않습니다.")
+            @Size(min = 8, max = 20, message = "아이디의 길이가 8-20글자 사이로 입력되지 않았습니다.") String directorId
+    ) {
+        GetDirectorResponse director = searchDiretorService.getDirector(directorId);
+        return new ResponseEntity<>(director, HttpStatus.OK);
+    }
+
+    @GetMapping("/searchDirector")
+    public ResponseEntity<List<SearchDirectorResponse>> searchDirector(@Valid @RequestBody SearchDirectorRequest request, @AuthenticationPrincipal String userIdByToken) {
+        List<SearchDirectorResponse> responsesByPaging = searchDiretorService.searchDirector(request, userIdByToken);
+        return new ResponseEntity<>(responsesByPaging, HttpStatus.OK);
+    }
```
위에 인증 관련 API들은 Restful 하기 어려운 API들이라서 이해가 되는데, 이런 API들도 Path가 이렇게 되어야할 이유가 있을까요..?
완성된 API의 Path가 어떤식이 될지, 은석님이 서비스에서 많이 보는 Path 형태일지 확인해보시면 좋을 것 같습니다. :)

tilsong replied:
@GetMapping("/director/{directorId}"), @PostMapping("/director/list") 으로 Path를 다시 생각해봤습니다. DirectorController를 따로 만들어서 둘까도 생각해봤는데, User 중 Director인 사람들을 조회하는 것이기 때문에 UserContoller 안에두었습니다.

f-lab-fi commented on src/main/java/com/directors/application/user/SearchDiretorService.java:
```diff
@@ -0,0 +1,193 @@
+
+        List<String> filteredIds = new ArrayList<>();
+
+        for (String id : userIds) {
+            boolean hasScheduleId = scheduleRepository
+                    .existsByUserIdAndScheduleStatus(id, ScheduleStatus.OPENED);
+            if (hasScheduleId) {
+                filteredIds.add(id);
+            }
+        }
```
이거 Stream 으로 작성해볼 수 있지 않을까요~?
그럼 가독성이 더 높을 것 같습니다.

tilsong replied:
넵 Stream으로 해보겠습니다!

f-lab-fi commented on src/main/java/com/directors/application/user/SearchDiretorService.java:
```diff
@@ -0,0 +1,193 @@
+            return userIds;
+        }
+
+        Set<String> filteredIds = new HashSet<>();
+
+        for (String id : userIds) {
+            addMatchingUserIds(searchText, filteredIds, id);
+            addMatchingSpecialtyUserIds(searchText, filteredIds, id);
+        }
```
이런 코드는 코드를 읽는 사람(특히 문제가 생겨서 디버깅 해야하는 사람)으로 하여금 filteredIds에게 무슨 일이 벌어지는지 확인하기 위해 addMatchingUserIds, addMatchingSpecialtyUserIds 메서드 안쪽으로 들어가도록 만듭니다.
filteredIds가 실제로 거기에서 요소에 추가되거나 제거되니까요.

가급적 이렇게 파라미터로 받아서 추가하기보다는 추가되어야할 요소가 반환되도록 만드시고, 호출하는 쪽에서 filteredIds에 요소를 추가하도록 작성하시는게 좋습니다.

f-lab-fi replied:
불가피한 상황도 분명 있긴한데, 이건 불가피한 상황은 아닌 것 같아요.

tilsong replied:
넵! 말씀하신 방향 대로 코드 작성해보겠습니다.

f-lab-fi commented on src/main/java/com/directors/application/user/SearchDiretorService.java:
```diff
@@ -0,0 +1,193 @@
+        List<UserRegion> userRegion = new ArrayList<>();
+
+        for (Address address : nearestAddress) {
+            userRegion.addAll(userRegionRepository.findByFullAddress(address.fullAddress()));
```
userRegion.addAll(userRegionRepository.findByFullAddress(address.fullAddress()));

한 행에 식이 3개나 됩니다.
가급적 행을 나눠주세요.

f-lab-fi commented on src/main/java/com/directors/application/user/SearchDiretorService.java:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    private List<String> filterUserIdBySpecialtyProperty(List<String> userIds, SpecialtyProperty specialtyProperty) {
+        List<String> filteredIds = new ArrayList<>();
+
+        for (String id : userIds) {
+            boolean hasSpecialtyPropertyId = specialtyRepository
+                    .findByUserId(id)
+                    .stream()
+                    .anyMatch(specialty -> specialty.getSpecialtyInfo().property().equals(specialtyProperty));
+
+            if (hasSpecialtyPropertyId) {
+                filteredIds.add(id);
+            }
+        }
```
이런 코드도 Stream으로 가독성 높은 코드가 될 수 있어보입니다. :)

tilsong committed:
[#20] 디렉터 목록 조회 기능 및 디렉터 상세 조회 기능 구현 - 불필요한 주석 제거

tilsong committed:
[#20] 디렉터 목록 조회 기능 및 디렉터 상세 조회 기능 구현 - 피드백 반영

- 디렉터 목록 조회 및 상세 조회 요청 path 변경

- 코드 가독성 있게 수정, 메소드를 호출하는 쪽으로 필요한 데이터가 반환될 수 있도록 수정.

tilsong committed:
Merge branch 'main' into feature/20

f-lab-fi commented on src/main/java/com/directors/application/user/SearchDiretorService.java:
```diff
@@ -89,7 +87,8 @@ private List<String> getNearestDirectorIds(String userId, int distance) {
         for (Address address : nearestAddress) {
-            userRegion.addAll(userRegionRepository.findByFullAddress(address.fullAddress()));
+            userRegion.addAll(
+                    userRegionRepository.findByFullAddress(address.fullAddress()));
```
아니 ㅋㅋㅋ
이렇게 나누라는 의미였을까요 ㅎ..
메서드 인자 안에서 또 다른 메서드를 호출하기보단 앞에서 지역변수로 받아서 사용하라는 의미입니다.
지금처럼 나누면 큰 의미가 없습니다.

tilsong committed:
[#20] 디렉터 목록 조회 기능 및 디렉터 상세 조회 기능 구현 - 피드백 반영2 및 코드 수정

- 중첩된 메서드 식 제거, 지역 변수 사용

- var 일부 적용

tilsong committed:
[#20] 디렉터 목록 조회 기능 및 디렉터 상세 조회 기능 구현 - 코드 스타일: var 적용

puallab reviewed (approved):
확인했습니다~!
